Rewrite iOS sleep timer to handle multi-track files

ronaldheft opened the pull request:
This took me quite a while to figure out, but I rewrote the iOS sleep timer again to address issues with multi-track books. Fixes #351.

The iOS sleep timer now:
* Uses a `Timer` for countdown-based sleep timers
* Uses a `TimeObserver` for a scheduled chapter end time

This seems to strike the balance of accuracy, reliability, and code readability. I have tested this PR with:
* Single-track books
* Multi-track books
* Podcasts

For each media type, I have verified:
* How the timer reacts to play/pause
* How the timer reacts to changes in playback speed
* How the timer reacts to seeking
* How the timer handles seeking past the end time (in case of a chapter end timer)

Why rewrite the sleep timer again? While the previous strategy worked, it involved recalculating the stop time each time seeking and playback rate changes occurred. Not to mention the edge cases with a chapter sleep time.

In theory, the previous strategy should work, but the way the player reinitialized when tracks are changed out during seeks with multi-file books caused the stop time to be recalculated wrong depending on the order of operations. I tried a number of different strategies to debug it, and race conditions would continue to come up, due to the fact that pause/stopping audio causes the playback rate to change on iOS. This would force the stop time to be recalculated, which would be occurring while the seek was happening. There is probably a way to solve for it, but I opted to simplify the logic.

ronaldheft commented on ios/App/Shared/player/AudioPlayer.swift:
```diff
@@ -378,32 +375,24 @@ class AudioPlayer: NSObject {
                 
                 if continuePlaying {
-                    self?.play()
+                    self.resumePlayback()
```
This change prevents a condition where seeking would occur within seeking, due to `play()` also needing to seek to the proper start time (even outside of `seekBack` conditions).

ronaldheft commented on ios/App/Shared/player/AudioPlayerSleepTimer.swift:
```diff
@@ -0,0 +1,168 @@
+import Foundation
+import AVFoundation
+
+extension AudioPlayer {
```
All the sleep timer logic is in one file now

advplyr commented:
Found one bug. After starting a sleep timer then changing the audiobook the sleep timer does not reset.

Then you have to cancel the sleep timer and start another to get it going again.

ronaldheft commented:
Good catch! I’m probably missing a call to cancel the timer on deinit. I’ll reproduce later this morning and verify that fixes the issue.

ronaldheft commented:
@advplyr Fixed it! Turned out `AudioPlayer` had a memory leak and `deinit` was not being called when switching player items.

Some of the `MPRemoteCommandCenter` targets had strong references to `self`, one of timers I had added had a strong reference, and the `AVAudioPlayer` observers also held strong references to `self`. FYI, I found [this article](https://stackoverflow.com/questions/41660486/knowing-where-retain-cycles-are-and-removing-them) really helpful in tracking down the bits of code holding onto references to the `AudioPlayer`.

Outside of fixing the memory leak, I also took a second look at the weird code I wrote to clear the chapter sleep timer when you skip back, and I rewrote that in a way that was not awkward.

advplyr commented:
Sleep timer is working perfectly in all my tests. Nice!
